src/Features/CSharp/Portable/CodeFixes/AddImport/CSharpAddImportCodeFixProvider.cs

```diff
@@ -546,65 +564,149 @@ private NameSyntax CreateNameSyntax(IReadOnlyList<string> namespaceParts, int in
+                return SyntaxFactory.ExternAliasDirective(SyntaxFactory.Identifier(externAliasString))
+                                    .WithAdditionalAnnotations(Formatter.Annotation);
             }
 
             return null;
         }
 
-        private UsingDirectiveSyntax GetUsingDirective(CompilationUnitSyntax root, INamespaceOrTypeSymbol namespaceSymbol, SemanticModel semanticModel, bool fullyQualify)
+        private UsingDirectiveSyntax TryGetUsingDirective(
+            INamespaceOrTypeSymbol namespaceOrTypeSymbol,
+            SemanticModel semanticModel,
+            CompilationUnitSyntax root,
+            SyntaxNode contextNode)
         {
-            if (namespaceSymbol is INamespaceSymbol)
+            var namespaceToAddTo = GetFirstContainingNamespaceWithUsings(contextNode);
+            var usingDirectives = namespaceToAddTo?.Usings ?? root.Usings;
+
+            var nameSyntax = namespaceOrTypeSymbol.GenerateNameSyntax();
+
+            // Replace the alias that GenerateTypeSyntax added if we want this to be looked
+            // up off of an extern alias.
+            var externAliasDirective = TryGetExternAliasDirective(
+                namespaceOrTypeSymbol, semanticModel, contextNode,
+                checkForExistingExternAlias: false);
+
+            var externAlias = externAliasDirective?.Identifier.ValueText;
+            if (externAlias != null)
             {
-                string namespaceString;
-                string externAliasString;
-                TryGetExternAliasString(namespaceSymbol, semanticModel, root, out externAliasString);
-                if (externAliasString != null)
-                {
-                    if (TryGetNamespaceString(namespaceSymbol, root, false, externAliasString, 
-                        checkForExistingUsing: true, namespaceString: out namespaceString))
-                    {
-                        return SyntaxFactory.UsingDirective(SyntaxFactory.ParseName(namespaceString).WithAdditionalAnnotations(Simplifier.Annotation));
-                    }
+                nameSyntax = AddOrReplaceAlias(nameSyntax, SyntaxFactory.IdentifierName(externAlias));
+            }
+            else
+            {
+                // The name we generated will have the global:: alias on it.  We only need
+                // that if the name of our symbol is actually ambiguous in this context.
+                // If so, keep global:: on it, otherwise remove it.
+                nameSyntax = RemoveGlobalAliasIfUnnecessary(semanticModel, nameSyntax, namespaceToAddTo);
+            }
 
-                    return null;
-                }
+            var usingDirective = SyntaxFactory.UsingDirective(nameSyntax)
+                                              .WithAdditionalAnnotations(Formatter.Annotation);
+
+            if (HasExistingUsingDirective(root, namespaceToAddTo, usingDirective))
```

Comment:
Rare engh that we don't bother trying to get that right.  Basically, our assumption is:

1. Nearly everyone will only have their usings either in the root, or in a nested namespace.
2. For people who have them in multiple scopes, we live with this case falling through.